model/loss/base_loss.py

```diff
@@ -557,3 +557,136 @@ def call(self, y_true: Dict[Text, tf.Tensor],
+    self._gt_key = gt_key
+    self._pred_key = pred_key
+
+  def call(self, y_true: Dict[Text, tf.Tensor],
+           y_pred: Dict[Text, tf.Tensor]) -> tf.Tensor:
+    """Computes the scale invariant logarithmic error.
+
+    Args:
+      y_true: A dict of tensors providing ground-truth information.
+      y_pred: A dict of tensors providing predictions.
+
+    Returns:
+      A tensor of shape [batch] containing the loss per sample.
+    """
+    gt = y_true[self._gt_key]
+    pred = y_pred[self._pred_key]
+
+    def _compute_error(loss_input):
+      gt, pred = loss_input
+      # Unlabeled pixels are filled with 0.
+      label_mask = gt > 0
+      gt = tf.boolean_mask(gt, label_mask)
+      pred = tf.boolean_mask(pred, label_mask)
+      # Scale invariant logarithmic error.
+      gt_log = tf.math.log(gt)
+      pred_log = tf.math.log(pred)
+      silog_error = (tf.reduce_mean(tf.square(gt_log - pred_log)) -
+                     tf.square(tf.reduce_mean(gt_log - pred_log)))
+      return silog_error
+
+    return tf.map_fn(_compute_error, (gt, pred), fn_output_signature=tf.float32)
+
+
+class RelativeSquaredError(tf.keras.losses.Loss):
+  """This class contains code to compute the relative squared error.
+
+  This class computes the relative squared error for monocular depth estimation.
+
+  Reference:
+  Uhrig, Jonas, Nick Schneider, Lukas Schneider, Uwe Franke, Thomas Brox, and
+  Andreas Geiger. "Sparsity invariant cnns." In 3DV, 2017.
+  """
+
+  def __init__(self,
+               gt_key: Text,
+               pred_key: Text):
+    # Implicit reduction might mess with tf.distribute.Strategy, hence we
+    # explicitly reduce the loss.
+    super().__init__(reduction=tf.keras.losses.Reduction.NONE)
+    self._gt_key = gt_key
+    self._pred_key = pred_key
+
+  def call(self, y_true: Dict[Text, tf.Tensor],
+           y_pred: Dict[Text, tf.Tensor]) -> tf.Tensor:
+    """Computes the relative squared error.
+
+    Args:
+      y_true: A dict of tensors providing ground-truth information.
+      y_pred: A dict of tensors providing predictions.
+
+    Returns:
+      A tensor of shape [batch] containing the loss per sample.
+    """
+    gt = y_true[self._gt_key]
+    pred = y_pred[self._pred_key]
+
+    def _compute_error(loss_input):
+      gt, pred = loss_input
+      # Unlabeled pixels are filled with 0.
+      label_mask = gt > 0
+      gt = tf.boolean_mask(gt, label_mask)
+      pred = tf.boolean_mask(pred, label_mask)
+      # Relative squared error.
+      relative_squared_error = tf.sqrt(
+          tf.reduce_mean(tf.square((gt - pred) / gt)))
+      return relative_squared_error
+
+    return tf.map_fn(_compute_error, (gt, pred), fn_output_signature=tf.float32)
+
+
+class SILogPlusRelativeSquaredLoss(tf.keras.losses.Loss):
+  """This class contains code to compute depth loss SILog + RelativeSquared.
+
+  This depth loss function combines the scale invariant logarithmic (SILog)
+  error and relative squared error.
```

Comment:
This depth loss function combines the scale invariant logarithmic (SILog) error and relative squared error, which was adopted in the ViP-DeepLab model.